### DISCUSSION TITLE
trait load_dsl is not implemented for a tuple of all sql_types

### DISCUSSION THREAD

**Discussion by Jeremiah-Griffin** [diesel-rs/diesel#3730] [Q&A]
I really don't know what this means or how to start fixing this.

Error (note that conn.mut_ref() returns an &mut SqliteConnection):
```
error[E0277]: the trait bound `(diesel::sql_types::Binary, diesel::sql_types::Text, diesel::sql_types::Double, d
iesel::sql_types::Double, diesel::sql_types::SmallInt, diesel::sql_types::BigInt, diesel::sql_types::Text): load
_dsl::private::CompatibleType<models::User, _>` is not satisfied
    --> src\types\mod.rs:136:28
     |
136  |             .first::<User>(conn.mut_ref())?
     |              -----         ^^^^^^^^^^^^^^ the trait `load_dsl::private::CompatibleType<models::User, _>` 
is not implemented for `(diesel::sql_types::Binary, diesel::sql_types::Text, diesel::sql_types::Double, diesel::
sql_types::Double, diesel::sql_types::SmallInt, diesel::sql_types::BigInt, diesel::sql_types::Text)`
     |              |
     |              required by a bound introduced by this call
     |
     = help: the following other types implement trait `load_dsl::private::CompatibleType<U, DB>`:
               (ST0,)
              *snip*
             and 24 others
     = note: required for `SelectStatement<FromClause<table>, DefaultSelectClause<...>, ..., ..., ..., ...>` to 
implement `LoadQuery<'_, _, models::User>`
     = note: the full type name has been written to 'C:\Users\J Griffin\Desktop\Ceres\thoth\target\debug\deps\th
oth-66407e051daf3315.long-type-14960283373290162428.txt'
note: required by a bound in `first`
    --> C:\Users\J Griffin\.cargo\registry\src\index.crates.io-6f17d22bba15001f\diesel-2.1.0\src\query_dsl\mod.r
s:1780:22
     |
1777 |     fn first<'query, U>(self, conn: &mut Conn) -> QueryResult<U>
     |        ----- required by a bound in this associated function
...
1780 |         Limit<Self>: LoadQuery<'query, Conn, U>,
     |                      ^^^^^^^^^^^^^^^^^^^^^^^^^^ required by this bound in `RunQueryDsl::first`

error[E0277]: the trait bound `models::User: FromSqlRow<_, _>` is not satisfied
    --> src\types\mod.rs:136:28
     |
136  |             .first::<User>(conn.mut_ref())?
     |              -----         ^^^^^^^^^^^^^^ the trait `FromSqlRow<_, _>` is not implemented for `models::Us
er`
     |              |
     |              required by a bound introduced by this call
     |
     = help: the following other types implement trait `FromSqlRow<ST, DB>`:
               <(T1, T0) as FromSqlRow<(ST1, Untyped), __DB>>
              *snip*
             and 23 others
     = note: required for `SelectStatement<FromClause<table>, DefaultSelectClause<...>, ..., ..., ..., ...>` to implement `LoadQuery<'_, _, models::User>`
     = note: the full type name has been written to 'C:\Users\J Griffin\Desktop\Ceres\thoth\target\debug\deps\thoth-66407e051daf3315.long-type-14960283373290162428.txt'
note: required by a bound in `first`
    --> C:\Users\J Griffin\.cargo\registry\src\index.crates.io-6f17d22bba15001f\diesel-2.1.0\src\query_dsl\mod.rs:1780:22


```

"User" database model:

```
#[derive(Queryable, Selectable, Insertable, Debug)]
#[diesel(table_name = crate::db::schema::users)]
#[diesel(check_for_backend(diesel::sqlite::Sqlite))]
pub struct User {
    #[diesel(serialize_as = Vec<u8>)]
    id: Uuid,
    pub address: String,
    latitude: f64,
    longitude: f64,
    pub country_code: i16,
    pub phone: i64,
    pub email: String,
}
```
Offending function:

```
 use diesel::*;
 use crate::db::schema::users::dsl::users;
 use crate::db::models::User;

let merchant_location  = users
            .find(merchant_id.as_bytes().to_vec())
            .first::<User>(conn.mut_ref())?
            .location();
```

**Comment by weiznich** [comment]
https://docs.diesel.rs/2.1.x/diesel/index.html#how-to-read-diesels-compile-time-error-messages

**Reply by Jeremiah-Griffin** [reply]
Yes, I've checked that page.

As you can see on the model, `Selectable` and `check_for_backend(diesel::sqlite::Sqlite)` are derived for `User` which is the suggestion given for this error as I understand it.

I still don't understand what I'm doing wrong.

**Reply by weiznich** [reply]
There is more on that page than `Selectable` + `check_for_backend(…)`.

>  This error indicates a mismatch between what your query returns and what your model struct expects the query to return. The fields need to match in terms of field order, field type and field count. If you are sure that everything matches, double check the enabled diesel features (for support for types from other crates) and double check (via cargo tree) that there is only one version of such a shared crate in your dependency tree. 

Given the provided information I cannot really add more here as I cannot check the field order (no schema provided), nor can I check the field types (It's unclear where the non-diesel, non-rust standard types are coming from), nor can I check the field count (again, no schema provided). If you would like to have further assistance you should try to provide a complete example of your problem.

As additional note: Your application of `#[derive(Selectable)]` is not complete. You would need to use `.select(User::as_select())` on your query to actually use the right select clause. This is [documented on the relevant trait](https://docs.diesel.rs/2.1.x/diesel/expression/trait.Selectable.html#examples)

**Reply by Jeremiah-Griffin** [reply]
all mentions of `users` in schema.rs:

```
diesel::table! {
    users (id) {
        id -> Binary,
        address -> Text,
        latitude -> Double,
        longitude -> Double,
        country_code -> SmallInt,
        phone -> BigInt,
        email -> Text,
    }
}

diesel::joinable!(cafes -> merchants (id));
diesel::joinable!(consumers -> users (id));
diesel::joinable!(grocers -> merchants (id));
diesel::joinable!(merchants -> users (id));

diesel::allow_tables_to_appear_in_same_query!(
    cafes,
    consumers,
    grocers,
    interested_users,
    merchants,
    transactions,
    users,
);
```

up.rs:

```
CREATE TABLE users(
    id                    BLOB NOT NULL UNIQUE PRIMARY KEY,
    address           TEXT NOT NULL,
    latitude           DOUBLE NOT NULL,
    longitude        DOUBLE NOT NULL,
    country_code  SMALLINT NOT NULL,
    phone             BIGINT NOT NULL,
    email               TEXT NOT NULL UNIQUE

) WITHOUT ROWID;
```

The `Uuid` is the from the `uuid` crate as used by the "uuid" feature. I'm not using that feature flag as I'm on the sqlite backend. I implemented `From<Uuid> for Vec<u8>` to get around that limitation.

cargo tree has only one mention of the uuid crate.

Here's the new function:

```
let merchant_location  = users
            .select(User::as_select())
            .find(merchant_id.as_bytes().to_vec())
            .first::<User>(conn.mut_ref())?
            .location();
```
Here's `.location()`:
```
 pub fn location(&self) -> geoutils::Location {
        Location::new_const(self.latitude, self.longitude)
    }
```

**Reply by weiznich** [reply]
> I'm not using that feature flag as I'm on the sqlite backend. I implemented From<Uuid> for Vec<u8> to get around that limitation.

This part is strange in so far that you cannot implement foreign traits (`From`) for foreign types `Uuid` and `Vec<u8>`. That means you likely don't use the `Uuid` type from the `uuid` crate there but a new type wrapper, right?
In addition to that above: Have you also implemented a `From<Vec<u8>> for Uuid` for your new type wrapper? If not that's likely your issue there (in combination with the missing `#[diesel(deserialize_as =… )]` attribute on the relevant field).

Now it would be interesting to see why there is no better error message with using `#[diesel(check_for_backend(…)]` but that's unfortunately impossible to debug as there is still missing information for that.

**Reply by Jeremiah-Griffin** [reply]
> This part is strange in so far that you cannot implement foreign traits (From) for foreign types Uuid and Vec<u8>. That means you likely don't use the Uuid type from the uuid crate there but a new type wrapper, right?

I forked Uuid. here's the commit, before today, this was the only way it differed from the release version of that crate.

https://github.com/Jeremiah-Griffin/uuid/commit/ee1bbf27dd78cb98a0018ae0762907a6eb775d6d

(in combination with the missing #[diesel(deserialize_as =… )] attribute on the relevant field).

[dirty Implementation here](https://github.com/uuid-rs/uuid/commit/1a8e8d43c5e9d3c5b15588d224c8429b9374078f).

New model:
```rust
#[derive(Queryable, Selectable, Insertable, Debug)]
#[diesel(table_name = crate::db::schema::users)]
#[diesel(check_for_backend(diesel::sqlite::Sqlite))]
pub struct User {
    #[diesel(serialize_as = Vec<u8>, deserialize_as = Uuid)]
    id: Uuid,
    pub address: String,
    latitude: f64,
    longitude: f64,
    pub country_code: i16,
    pub phone: i64,
    pub email: String,
}
```

Same errors.

>Now it would be interesting to see why there is no better error message with using #[diesel(check_for_backend(…)] but that's unfortunately impossible to debug as there is still missing information for that.

Please tell me what you need. This is my first time using diesel and I don't know what else I need to provide to you. 

Additionally, is there somewhere where the use of custom types with diesel is documented? I could not find much on the docs.rs or the diesel.rs site. What I have done here is entirely the product of searching stack overflow and grepping through the diesel repo.

**Reply by Jeremiah-Griffin** [reply]
Realizing that I was overfetching and I needed to extract this behavior to a new function I refactored:

original: 
```rust
let merchant_location  = users
            .select(User::as_select())
            .find(merchant_id.as_bytes().to_vec())
            .first::<User>(conn.mut_ref())?
            .location();
```

new:
```rust
 let (lat, lon) = users
        .find(user_id.as_bytes().to_vec())
        .select((latitude, longitude))
        .first::<(f64, f64)>(conn.mut_ref())?;

    Ok(Location::new_const(lat, lon))
```

The refactored function compiles.

I still need assistance with the original matter however, as eventually i'll need to retrieve whole structs from the db as I was doing before.

**Reply by weiznich** [reply]
> Please tell me what you need. This is my first time using diesel and I don't know what else I need to provide to you.

To provide further help you need to provide a complete minimal example (== something that reproduces the error if I run `cargo check` and that contains nothing that's not required for reproducing the error).

> Additionally, is there somewhere where the use of custom types with diesel is documented?

There is quite a bit documentation about that topic in our API documentation in the documentation of the relevant traits/derives. See [here for `FromSql`](https://docs.diesel.rs/2.1.x/diesel/deserialize/trait.FromSql.html#examples) and [here for `ToSql`](https://docs.diesel.rs/2.1.x/diesel/serialize/trait.ToSql.html#examples) and [here for `Queryable`](https://docs.diesel.rs/2.1.x/diesel/deserialize/trait.Queryable.html#deserialize_as-attribute) and [here for `Insertable`](https://docs.diesel.rs/2.1.x/diesel/prelude/derive.Insertable.html#examples). Additionally there is an [complete example](https://github.com/diesel-rs/diesel/tree/master/examples/postgres/custom_types) and a [draft guide in the repo](https://github.com/diesel-rs/diesel/blob/master/guide_drafts/custom_types.md)
